pipeline.py

```diff
@@ -1,83 +1,125 @@
+
+    # Create Model RDM
+    # just in case different subjects have different label data.
+    # If not, then this could be done just once.
+    print("Creating model RDM")
+    model_RDM = create_model_RDM(labels_sorted.labels.to_numpy())
+    model_RDM_pic = matshow(model_RDM)
+    savefig(f"model_RDM_subject{subj_no}.png")
+
+    # create voxel index
+    print("Creating voxel index...")
+    data_shape = voxels_sorted.shape
+    voxel_shape = data_shape[0:3]
+    voxel_index_grid = list(np.ndindex(voxel_shape))
+
+
+    # Create list of ROI voxel locations:
+    ROI_locs = list(map(tuple, np.argwhere(mask_data == 1)))
+
+
+    # create BOLD RDMs for all center voxels:
+    ROI_RSAs  = {}
+    ROI_RDM_progress_bar = Bar('Computing RDMs for ROIs', max=len(ROI_locs))
+
+    # create a searchlight generator for returning search locations per single center voxel
+    searchlight_gen = searchlight_generator(ROI_locs, searchlight_radius, voxel_index_grid)
+
+    for loc in ROI_locs:
+        searchlight_locs = next(searchlight_gen)
+        ROI_RDM = create_bold_RDM(voxels_sorted, searchlight_locs)
+        ROI_RSAs[loc], RSAp = spearmanr(ROI_RDM.flatten(), model_RDM.flatten())
+        if np.isnan(ROI_RSAs[loc]):
+            print(ROI_RDM)
+            raise ValueError("RSA value was nan")
+            break
+        ROI_RDM_progress_bar.next()
+
+    ROI_RDM_progress_bar.finish()
 
-data = bold.get_fdata()
-mask_data = mask.get_fdata()
 
-# Preprocessing
-print("Pre-processing...")
-chunk_id = labels.chunks.to_numpy()
-processed_bold_data = preprocess_bold_data(data, chunk_id)
-preprocessed_image = nib.Nifti1Image(processed_bold_data, bold.affine)
+    # create an RSA image:
+    print("Creating a Nifti1Image")
+    plot_data = np.zeros(data.shape[0:3])
+    for ROI_loc in ROI_RSAs:
+        plot_data[ROI_loc] = ROI_RSAs[ROI_loc]
 
-# Remove unwanted images from data
-pic_voxels, pic_labels = remove_rows(processed_bold_data, labels, ['rest', 'scrambled pix'])
+    print("Plotting RSA figures")
+    img = nib.Nifti1Image(plot_data, bold.affine)
+    plotting.plot_glass_brain(img, output_file=f"RSA_plot_glass_brain_subj{subj_no}.png")
 
-print("Sorting...")
-labels_sorted = pic_labels.sort_values('labels')
-voxels_sorted = pic_voxels[:, :, :, pic_labels.index]
+    return plot_data
 
-# create voxel index
-print("Creating voxel index...")
-data_shape = voxels_sorted.shape
-voxel_shape = data_shape[0:3]
-voxel_index_grid = list(np.ndindex(voxel_shape))
+##########################
+# Multi-subject pipeline #
+##########################
 
 
-# Create Model RDM
-print("Creating model RDM")
-model_RDM = create_model_RDM(labels_sorted.labels.to_numpy())
-model_RDM_pic = matshow(model_RDM)
-savefig("model_RDM.png")
+# finds out the shape of data based on first subject in list
+# assumes all subjects have same shape
+first_bold = nib.load(f'data/subj{subject_list[0]}/bold.nii.gz')
+bold_data_shape = first_bold.get_fdata().shape[0:3]
```

Comment:
Some premature optimization here. The RSA result is only the size of a single MRI image. You can easily just store a list of them and call `np.mean`. No need for these shenanigans of summing them in-place.